(PUP-9930) Conditionally load openssl when not in JRuby FIPS

justinstoller opened the pull request:
Just putting this up for discussion, needs a ticket, cleanup, maybe totally different approach even....

justinstoller commented on lib/puppet/ssl/ssl_provider.rb:
```diff
@@ -148,6 +148,13 @@ def default_flags
   end
 
   def create_x509_store(roots, crls, revocation)
+    # TODO: This method is called in one Server test. (See:
```
This is the test in question  https://github.com/puppetlabs/puppetserver/blob/6.5.0/test/integration/puppetlabs/services/master/master_service_test.clj#L664-L699 and I'm not sure yet whether the test is bad, Puppet's behavior is bad, or what.

joshcooper commented on lib/puppet/util/openssl_loader.rb:
```diff
@@ -0,0 +1,95 @@
+
```
nit could you move this to `lib/puppet/ssl/openssl_loader.rb`?

justinstoller replied:
absolutely

puppetcla commented:
CLA signed by all contributors.

justinstoller commented on lib/puppet/util/monkey_patches.rb:
```diff
@@ -28,34 +28,35 @@ def daemonize
-    DEFAULT_PARAMS[:ciphers] << ':!SSLv2'
-  end
+require 'puppet/ssl/openssl_loader'
+if !defined?(JRUBY_VERSION)
```
Should use the `Puppet::Util::Platform.jruby_fips?` introduced elsewhere in the PR

justinstoller commented on lib/puppet/util/platform.rb:
```diff
@@ -33,6 +33,22 @@ def fips_enabled?
+                           false
+                         end
+      end
+      module_function :jruby_fips?
```
Is this check sensible?

Magisus replied:
Seems simple enough, if there isn't a more obvious check (like a class that is only in FIPS and is named accordingly). But it seems like generally speaking the FIPS stuff is a subset of regular BC?

justinstoller replied:
That's been my impression. The lack of this class is the first thing you see when trying to load jopenssl in FIPS.

justinstoller commented:
I'm not clear, should I also refactor the spec tests?

Magisus commented:
Also, gonna vote no on fixing the specs. As far as I know, we don't plan to run the unit tests under JRuby in FIPS mode (seems like low ROI there), so it doesn't seem worth the amount of work it would be to get them working.

If someone knows of a plan to do this, please correct me.

justinstoller commented:
So the failures in Travis are unexpected... On MRI it's failing with:
```
  1) Integration Tests for Puppet::Type::Service::Provider::Systemd should be considered suitable if /bin/systemctl is present
     Failure/Error: expect(provider_class).to be_suitable
       expected `Puppet::Type::Service::ProviderSystemd.suitable?` to return true, got false
     # ./spec/integration/provider/service/systemd_spec.rb:11:in `block (2 levels) in <top (required)>'
     # util/rspec_runner:44:in `run'
     # util/rspec_runner:59:in `<main>'
Finished in 22.49 seconds (files took 5.8 seconds to load)

Finished in 3 minutes 23.5 seconds
21738 examples, 1 failure, 39 pending
```
Which perhaps unsurprisingly doesn't fail on my Mac (all specs pass for me in MRI). I'm confused if this even _could_ be related to my changes.

The JRuby specs fail with:
```
$ gem update --system && gem install bundler --no-document
Picked up _JAVA_OPTIONS: -Xmx1024m -Xms512m
Error occurred during initialization of boot layer
java.lang.module.FindException: Unable to derive module descriptor for /home/travis/.rvm/rubies/jruby-9.2.0.0/lib/jruby.jar
Caused by: java.lang.module.InvalidModuleDescriptorException: Test.class found in top-level directory (unnamed package not allowed in module)
The command "gem update --system && gem install bundler --no-document" failed and exited with 1 during .
```
Which makes no sense to me and sounds like a problem with Travis' infrastructure. However, locally running specs using JRuby 9.2.4.0 I get:
```
Finished in 2 minutes 22.8 seconds (files took 21.9 seconds to load)
Finished in 9 minutes 38 seconds
20644 examples, 263 failures, 63 pending
```
Which, unfortunately, is the same failures that I get when running _without_ this patch in my local environment:
```
Finished in 2 minutes 17.8 seconds (files took 23.2 seconds to load)
Finished in 9 minutes 42 seconds
20644 examples, 263 failures, 63 pending
```

justinstoller commented:
I'm a little lost on the travis failures if someone from @puppetlabs/platform-core could help me out.

kris-bosland commented:
@justinstoller , it may be related to Travis updates, my only info comes from comments here: https://github.com/puppetlabs/puppet/pull/7649

justinstoller commented:
Rebased onto HEAD of master. Also updated the connection ssl_context validation/warnings to be guarded against running on JRuby. I think this is ready to go now, assuming the Travis gods are with me.

Locally I get:
```
 + c4c926b...d39d7fb   fipsy -> fipsy (forced update)
11:50:29 sellout:puppet justin (fipsy $%) ‖ be rake parallel:spec
Processing 22 spec group(s) with 8 worker(s)
..............................................................................
...snip....
....
Finished in 1 minute 44.75 seconds
21743 examples, 0 failures, 39 pending

11:53:37 sellout:puppet justin (fipsy $%) ‖ ruby --version
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin18]
```

justinstoller commented:
ping @puppetlabs/team-coremunity , I'd love a review on this when you have a chance.

joshcooper commented on lib/puppet/network/http_pool.rb:
```diff
@@ -73,21 +74,25 @@ def self.http_ssl_instance(host, port, verifier = Puppet::SSL::Validator.default
-      end
+    # When running in a JRuby master environment the Server manages our
+    # SSLContext for us.
+    if !Puppet::Util::Platform.jruby?
```
The diff is a little confusing, but I think this change means the `connection` method is a noop on JRuby? If so, that will break puppetserver acting as an HTTPS client when submitting facts, etc to puppetdb. I'd expect the method to still call `http_client_class.new(host, port, ...)` where it calls the class that puppetserver sets.

justinstoller replied:
Yeah, that's ridiculous. I meant to condition the check/warning/error on JRuby's presence, not the whole creation of the client.

joshcooper commented on spec/unit/network/http_pool_spec.rb:
```diff
@@ -49,7 +50,7 @@ def initialize(host, port, options = {})
     end
 
-    context "when calling 'connection'" do
+    context "when calling 'connection'", :unless => Puppet::Util::Platform.jruby? do
```
Related to comment above, I'd expect to run these tests on JRuby

justinstoller replied:
I've removed the conditional from the overall context and moved it down to the examples that test the error/warnings.

justinstoller commented:
Updated

joshcooper commented on lib/puppet/network/http_pool.rb:
```diff
@@ -73,16 +74,18 @@ def self.http_ssl_instance(host, port, verifier = Puppet::SSL::Validator.default
+    # SSLContext for us.
     if use_ssl
-      unless ssl_context
+      if !ssl_context && !Puppet::Util::Platform.jruby?
```
If we allow a nil ssl_context then other code needs to account for it possibly being nil like the `Verifier` below. Could we instead have puppetserver create a dummy/empty context and push that on the stack? Then we wouldn't need to make this conditional on jruby.

justinstoller replied:
In a separate thread about trust anchors we talked about the SSLContext being a thing that JRuby interrogated to find out what certs to load for its http client. With the caveat that the SSLContext it interrogated mustn't load OpenSSL for FIPS in JRuby compatibility.

This current work seems like it would be superseded by that other SSLContext work. eg, this is difficult because the Server's http client class doesn't have the same constructor API as Puppet's Ruby client wrt SSLContext & Verifier (what this method is concerned with). Maybe instead of creating an empty context, it'd be better to create that non-OpenSSL context that Server can interrogate for certs, then update the Server's http client class to honor that context in its constructor?

Or, going in the opposite direction, could we move these checks that happen in this method into the client class (it becomes the responsibility of the Ruby or JRuby client classes to find and validate their ssl contexts)?

I might be way off base on the connection between these two issues....

justinstoller replied:
tl;dr was that I was off base

joshcooper commented on lib/puppet/ssl/openssl_loader.rb:
```diff
@@ -0,0 +1,24 @@
+      class Request; end
+      class Certificate; end
+    end
+  end
```
When puppet is required the cert/key indirections are loaded. I'm kind of amazed that works given the empty definitions above. Puppet should no longer call those indirections for anything normally, but I'm curious what would happen if a report processor or puppetdb terminus were to call `Puppet::SSL::Host.localhost`, which will trigger the respective termini. Should we be more defensive about that method and possibly setting the cert/key termini to `:none` or some stupid thing like that? I'm fine if the answer is no, just curious how things might fail.

justinstoller replied:
:mindblown:

justinstoller replied:
I would assume any call to `Host.localhost` would fail with an undefined constant exception. I'll look into setting the termini to `:none` though I'm not sure I understand why someone would use `.localhost` who wasn't needing to bootstrap Puppet's files.

joshcooper commented on lib/puppet/util/platform.rb:
```diff
@@ -33,6 +33,22 @@ def fips_enabled?
+
+                           begin
+                             require 'openssl'
+                             false
```
In MRI ruby you can ask openssl if fips is enabled:

```
$ ruby -ropenssl -e "puts OpenSSL::OPENSSL_FIPS"
false
```

Should we do the same for JRuby to future-proof jruby+fips being able to load openssl?

justinstoller replied:
You can ask JRuby's OpenSSL implementation if it's fips enabled (it's hardcoded to false), but the way we've done our FIPS builds you can't load it's OpenSSL to ask it that question.

joshcooper replied:
😆 😭

joshcooper commented:
jenkins please test this
